### PR TITLE
Support batch for LKJ transform

### PR DESCRIPTION
**Pull request by fehiepsi**
@elbamos I vectorized the inverse function, fixed the bug (from Stan) for derivative of tanh, and support batch for all transform in this PR. Hope that it does not duplicate your work.

Our next target will be the distribution. Would you want to revise it first or I'll do?

### CODE REVIEW

File: pyro/distributions/lkj.py [hunk 1]
```diff
@@ -62,65 +60,55 @@ class _PartialCorrToCorrCholeskyTransform(Transform):
 
     Reference:
 
-    [1] `Generating random correlation matrices based on vines and extended onion method`,
-    Daniel Lewandowski, Dorota Kurowicka, Harry Joe
+    [1] `Cholesky Factors of Correlation Matrices`,
+    Stan Reference Manual v2.18, Section 10.12
```
Comment (fehiepsi) [inline pyro/distributions/lkj.py:64]: Lewandowski does not mention about this transform (only transform from PartialCorr to Corr) so Stan Reference Manual will be our main reference.

**Comment by elbamos** [comment]
I've already got the distribution; if the batched transformation checks-out the distribution should be fine. I'll go through this tomorrow.

File: pyro/distributions/lkj.py [hunk 2]
```diff
@@ -49,10 +47,10 @@ def _call(self, x):
         return x.tanh()
 
     def _inverse(self, y):
-        return torch.log((1 + y) / (1 - y)) / 2
+        return ((1 + y) / (1 - y)).log() / 2
 
     def log_abs_det_jacobian(self, x, y):
-        return (-2) * y.cosh().log()
+        return (1 - y ** 2).log()
```
Comment (fehiepsi) [inline pyro/distributions/lkj.py:53]: this is a bug in Stan manual!

File: pyro/distributions/lkj.py [hunk 3]
```diff
@@ -62,65 +60,55 @@ class _PartialCorrToCorrCholeskyTransform(Transform):
...
+            pos_x = new_pos_x
+        return y
 
-    def log_abs_det_jacobian(self, x, z):
-        return (1 - x.tril(-1).pow(2).sum(1)).log().sum() * .5
+    def _inverse(self, y):
+        x = y.new_ones(y.shape)
+        triu_index = x.triu(diagonal=1) > 0.5
+        x[..., :, 0] = y[..., :, 0]
+        x[..., :, 1:] = y[..., :, 1:] / (1 - y.pow(2).cumsum(-1)[..., :, :-1]).sqrt()
```
Comment (fehiepsi) [inline pyro/distributions/lkj.py:102]: although the computation for upper triangular is wasting, this version is still much faster than loop version (for both small D and large D)

File: pyro/distributions/lkj.py (hunk 2 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:53]: This isn't a bug in the stan manual.  `1/cosh(y)^2 == 1 - tanh(y)^2`. 

The log-absolute-determinant has to be multiplied across the elements in the batch.  Its more numerically stable to do it the Stan way, which is `-2 * y.cosh().log().sum(some axis)`

File: pyro/distributions/lkj.py [hunk 4]
```diff
@@ -21,10 +21,8 @@ class _CorrCholesky(Constraint):
     Euclidean norm of each row is 1.
     """
     def check(self, value):
-        if not lower_cholesky().check(value): return False
-
         unit_norm_row = ((value.pow(2).sum(-1) - 1).abs() < 1e-6).min(-1)[0]
-        return unit_norm_row
+        return constraints.lower_cholesky.check(value) & unit_norm_row
```
Comment (elbamos) [inline pyro/distributions/lkj.py:25]: Is the purpose of this change to support batching?

File: pyro/distributions/lkj.py (hunk 2 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:53]: Not sure why my comment on this isn't showing up...

Anyway it isn't a bug in the stan manual.  `1/cosh(y)^2 == 1 - tanh(y)^2`.  But you have to take the product of the derivative across all of the elements in the batch.  Its simpler and more numerically stable to do that the stan way, which is `-2 * y.cosh().log().sum(some axis)`.

Comment (fehiepsi) [reply pyro/distributions/lkj.py:53]: I think that `y**2` is more numerically stable than `x.cosh()`.

About testing, could you suggest some else tests which I missed? Everything looks fine according to the current tests (including batch, scalar, singleton,...)

Comment (fehiepsi) [reply pyro/distributions/lkj.py:53]: We don't need to take the sum. ComposedTransform does it automatically if we specify a correct `event_dim`. (you can verify it by yourself or looking at the test for the composed transform).

File: pyro/distributions/lkj.py (hunk 4 above)
Comment (fehiepsi) [reply pyro/distributions/lkj.py:25]: Yes, and to correct the last PR, which returns True/False instead of Bool Tensor.

**Comment by fehiepsi** [comment]
@elbamos I have added tests for these transforms pretty throughout. Please let me know if I missed some tests for edge cases. Thanks!

File: pyro/distributions/lkj.py (hunk 2 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:53]: I may be mistaken about this --- but in thinking about it, its hard to see a way to get the calculation right when tanh is a separate transform. It looks to me that the solution is to go back to the way it was setup before, where the transformation from reals->cholesky is a subclass of the transformation from [-1,1]->cholesky, so that we know the event dimension when lob_abs_det_jacobian gets called.  

If you want to write a reals->[-1,1] transformation using tanh (and I'm not sure why you would want to do this -- the usual way to enforce a range transformation is with a modified logit), then I think that should take place separately from this project. 

Anyway, let's try to deal with changes to the tanh part of the code separately from the changes to the cholesky transformation, in a separate PR.  OK?

File: pyro/distributions/lkj.py (hunk 4 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:25]: The reason I'm asking is that if the `lower_cholesky` check fails, I'm not sure why we'd put in the work to do the `unit_norm_row` check. 

Also -- I'm pretty sure there are faster ways to do this.  One would be to call `value.norm(dim=-1)`. (The alternative would be to do take the diagonal of the `crossprod`.)  No?

Also - do we need the eps here?  Shouldn't the diagonal be exactly 1.0?

File: pyro/distributions/lkj.py [hunk 5]
```diff
@@ -62,65 +60,55 @@ class _PartialCorrToCorrCholeskyTransform(Transform):
...
+        x = y.new_ones(y.shape)
+        triu_index = x.triu(diagonal=1) > 0.5
+        x[..., :, 0] = y[..., :, 0]
+        x[..., :, 1:] = y[..., :, 1:] / (1 - y.pow(2).cumsum(-1)[..., :, :-1]).sqrt()
+        # we transpose and take upper triangular indices to arrange the result vector
+        # by (x21, x31, x41,..., x32, x42,...) instead of (x21, x31, x32, x41, x42,...)
+        return x.transpose(-1, -2)[triu_index].reshape(x.shape[:-2] + (-1,))
+
+    def log_abs_det_jacobian(self, x, y):
+        triu_index = y.new_ones(y.shape).triu(diagonal=1) > 0.5
```
Comment (elbamos) [inline pyro/distributions/lkj.py:108]: Is `.triu` going to operate correctly when there's one or more dimensions to the batch? I.e., if the input has shape `[4, 2, 3, 3]`, is `.triu` going to do what's expected? 

This approach with the indices also looks to be slower than the approach of just taking the triangle, flattening it, and summing it. 

Also, the transformation should be making the *lower* cholesky factor, not the upper one.  The reason is that the `multinormal` and related distributions all take a `scale_tril` parameter, which is the *lower* cholesky factor of the covariance matrix.

File: pyro/distributions/lkj.py (hunk 3 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:102]: What range of D values did you test this on?

**Comment by elbamos** [comment]
I've made a number of comments on different parts of this.  I haven't tried to pull it and test it.  

I would appreciate if you could address the comments, and split this up into separate PRs. One for the basic transform, one for the constraint, one for the log abs determinant, one for the tanh part of the transform.  Its pretty hard to review when its this monolithic.

File: pyro/distributions/lkj.py (hunk 2 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:53]: Ok I see your response now. If you let ComposedTransform take care of it, then aren't you getting a less efficient and less numerically stable gradient?

File: pyro/distributions/lkj.py (hunk 3 above)
Comment (fehiepsi) [reply pyro/distributions/lkj.py:102]: D=3, D=10, D=100

File: pyro/distributions/lkj.py (hunk 2 above)
Comment (fehiepsi) [reply pyro/distributions/lkj.py:53]: If you think something is more stable, then could you please add a test? I can't identify what causes "less numerically stable gradient" here. When I added tests for your code, it didn't pass, so I can't compare. So what I take from your argument is: Stan guys don't do Tanh transformation as in their code?

Comment (elbamos) [reply pyro/distributions/lkj.py:53]: No, that's not the issue.  Your implementation of it effectively (because you're relying on the ComposeTransform to do some of the work), is `(1 - tanh(x)^2).log().sum()`.  The other approach is `(-2 * cosh(x).log().sum())`.  See the difference? 

(Also, does this function take as its argument the pre-transformed value, or the post-transformed value?)

File: pyro/distributions/lkj.py (hunk 4 above)
Comment (fehiepsi) [reply pyro/distributions/lkj.py:25]: This method returns Bool Tensor (which shape maybe (1000, 1000, 1000)). We can't do if/else for it.

`value.norm()` is fine for small dimension tensor. For large dimension tensor, avoiding taking square root is faster (you can test it if you want).

We need eps due to numerical precision, we can't compare something to an exact value.

File: pyro/distributions/lkj.py (hunk 5 above)
Comment (fehiepsi) [reply pyro/distributions/lkj.py:108]: Yes, triu support batch. You can see the reason in my comment about why using triu_index in the code.

Comment (elbamos) [reply pyro/distributions/lkj.py:108]: I saw it, but I wasn't persuaded.  Also, do you see the point about needing to return a *lower* cholesky factor?

**Comment by fehiepsi** [comment]
@elbamos I don't think that there are much to review. If you worry about the correctness, then please tell me what else you want to test. I have added tests for batch, singleton, scalar,... About the effectiveness, if you come up with a better version, please let me know. I just vectorize your implementation and fix some bugs.

I can't split this PR because your code does not work! I have to fix them. If not, then tests can't pass. I don't know why you don't add tests in the first place. And I don't know what you have tested (as you claimed) because I can't see it here.

**Comment by elbamos** [comment]
A few days ago, you asked me to take a different approach to the way I pushed code. I've respected that, and for that reason I haven't pushed any updates, including code that fixed the issues you identified and implemented batching. 

I would appreciate if you would show me the same courtesy by breaking this up into separate PRs for the different issues. 

1. The constraint should be its own PR.  I'm fine with the changes you made here.

2. The tanh part of the transformation should be its own PR. It should not try to implement a standalone real->[-1,1] transformation. That transformation is more appropriately handled through other means. It should instead simply subclass the main transformation and apply the more numerically stable version of the gradient.

3. The change to the main transformation should be its own PR. It needs to produce a lower rather than an upper cholesky factor.  I have some other concerns about the approach. Maybe those concerns are valid, maybe not. But its very difficult to get into them in detail when its all together like this.

File: pyro/distributions/lkj.py (hunk 2 above)
Comment (fehiepsi) [reply pyro/distributions/lkj.py:53]: (I have written template with links to pytorch's transforms for you. You can check it to see what x, y play. Here x is in domain, and y is in codomain.)

I don't see what you mean. Could you elaborate on it? Assume we don't separate but combine two transforms into one class, then will you compute `tanh(x)` in log_abs_det_jabobian? If yes, then why we don't use it but computing cosh?

File: pyro/distributions/lkj.py (hunk 5 above)
Comment (fehiepsi) [reply pyro/distributions/lkj.py:108]: I don't see what you mean, could you elaborate it, pls?

**Comment by fehiepsi** [comment]
The thing is you didn't test your changes. So I suggest you making PRs to not wasting our time. This PR only supports batch and add tests.

About Tanh transformation, maybe you have another idea. So I'll close this PR and let it to you to update your code, fix bugs, add tests first.

Note that I transpose before taking upper indices, so there is no upper Cholesky is involved here.

File: pyro/distributions/lkj.py (hunk 2 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:53]: Your approach to doing it applies more operations, and does so to more components of the tensor, without any observable benefit.  

While `x**2` is more stable than `cosh(x)`, `1 - tanh(x)**2` is not more stable than `cosh(x)`. 

I'm not saying to combine the two transforms into one class.  I'm saying the transformation from `real->lowercholesky` can subclass the transformation from `[-1,1]->lowercholesky`. (Or it could be a separate transform that then calls `[-1,1]->lowercholesky` internally, that would be fine too.)

Comment (fehiepsi) [reply pyro/distributions/lkj.py:53]: Because I can't understand, so it is better to see your implementation instead.

File: pyro/distributions/lkj.py (hunk 5 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:108]: I mean the transformation needs to be returning a cholesky factor in the *lower* triangle of the matrix, not the *upper* triangle. Because the result is going to be fed into distribution classes that take as a parameter the *lower* cholesky factor, not the *upper* one.

**Comment by elbamos** [comment]
Ok.  I will try to use as much of your code as I can.

**Comment by fehiepsi** [comment]
You don’t have to do so. Just make sure it work first (by using tests). I
can do following PRs in pyro to address the effectiveness.

File: pyro/distributions/lkj.py (hunk 5 above)
Comment (elbamos) [reply pyro/distributions/lkj.py:108]: BTW - I'm working on this now, and I'm trying to use/test your code for the inverse.  I'm having trouble with the call to `triu` not supporting batches:

```
>>> torch.randn(2, 3, 3).triu(diagonal=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: invalid argument 1: expected a matrix at /Volumes/ShareData/LIBRARIES/pytorch/aten/src/TH/generic/THTensorMoreMath.cpp:1286
```

Am I missing something?  And would you prefer if I do a push where I leave inverse unimplemented so you can do it?

Comment (fehiepsi) [reply pyro/distributions/lkj.py:108]: You would need to use pytorch 1.0.1 or nightly I guess

Comment (fehiepsi) [reply pyro/distributions/lkj.py:108]: I guess the reason you were worried about the performance/optimized code instead of focusing on correctness first is because you have your own version work. So just use your version. It’ll reduce your time. I’ll not make PRs to this branch anymore because I can’t follow your idea and I think it is ready (after you fix your errors and add your tests). I’ll optimize the code in the following PRs in pyro repository, where I can give benchmark examples to compare, rather than to discuss about effectiveness or ineffectiveness “without code” here.
